go/signedexchange/cbor/encoder_test.go

```diff
@@ -41,15 +44,145 @@ func TestIntegers(t *testing.T) {
 
-		if !bytes.Equal(expected, b.Bytes()) {
-			t.Errorf("%d expected to encode to %v, actual %v", test.i, expected, b.Bytes())
+		if !bytes.Equal(exp, b.Bytes()) {
+			t.Errorf("%v expected to encode to %v, actual %v", test.bs, exp, b.Bytes())
 		}
 	}
 }
+
+func TestEncodeTextString(t *testing.T) {
+	var textTests = []struct {
+		s        string
+		encoding string
+	}{
+		{"", "60"},
+		{"a", "6161"},
+		{"IETF", "6449455446"},
+		{`"\`, "62225c"},
+		{"\u00fc", "62c3bc"},
+		{"\u6c34", "63e6b0b4"},
+		{"\U00010151", "64f0908591"},
+	}
+	for _, test := range textTests {
+		var b bytes.Buffer
+		e := NewEncoder(&b)
+
+		if err := e.EncodeTextString(test.s); err != nil {
+			t.Errorf("Encode. err: %v", err)
+		}
+		exp := fromHex(test.encoding)
+
+		if !bytes.Equal(exp, b.Bytes()) {
+			t.Errorf("\"%s\" expected to encode to %v, actual %v", test.s, exp, b.Bytes())
+		}
+	}
+}
+
+func TestMapEncoderSorter(t *testing.T) {
+	// The keys in every map MUST be sorted in the bytewise lexicographic order of
+	// their canonical encodings. For example, the following keys are correctly
+	// sorted:
+	// 1. 10, encoded as 0A.
+	// 2. 100, encoded as 18 64.
+	// 3. -1, encoded as 20.
+	// 4. “z”, encoded as 61 7A.
+	// 5. “aa”, encoded as 62 61 61.
+	// 6. [100], encoded as 81 18 64.
+	// 7. [-1], encoded as 81 20.
+	// 8. false, encoded as F4.
+	es := []*MapEntryEncoder{
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeInt(10)
+			valueE.EncodeTextString("int 10")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeInt(100)
+			valueE.EncodeTextString("int 100")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeInt(-1)
+			valueE.EncodeTextString("int -1")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeTextString("z")
+			valueE.EncodeTextString("string \"z\"")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeTextString("aa")
+			valueE.EncodeTextString("string \"aa\"")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeArrayHeader(1)
+			keyE.EncodeInt(100)
+			valueE.EncodeTextString("array [100]")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeArrayHeader(1)
+			keyE.EncodeInt(-1)
+			valueE.EncodeTextString("array [-1]")
+		}),
+		GenerateMapEntry(func(keyE *Encoder, valueE *Encoder) {
+			keyE.EncodeBool(false)
+			valueE.EncodeTextString("false")
+		}),
+	}
+
+	keyExp := [][]byte{
+		fromHex("0A"),
+		fromHex("18 64"),
+		fromHex("20"),
+		fromHex("61 7A"),
+		fromHex("62 61 61"),
+		fromHex("81 18 64"),
+		fromHex("81 20"),
+		fromHex("F4"),
+	}
+	for i, exp := range keyExp {
+		if !bytes.Equal(exp, es[i].Bytes()) {
+			t.Errorf("FAIL key %d expected to encode to %v, actual %v ", i, exp, es[i].Bytes())
+		}
+	}
+
+	if !sort.SliceIsSorted(es, func(i, j int) bool {
+		return bytes.Compare(es[i].Bytes(), es[j].Bytes()) < 0
```

Comment:
please put the compare func in cbor package, since this corresponds to the [Canonical CBOR](https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#rfc.section.3.5) concept

Reply:
As we discussed offline, let's not put sorting function as a public API in cbor.